Jetpack UX: Add xml and rest api error handling

kozer opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/dotcom-forge/issues/3241

## Proposed Changes

This pr add support for jetpack XML-RPC and REST API connection is

## Testing Instructions

### Test 1: REST API errors
1. Apply D119857-code
2. Create a new WoA site.
3. Enable ssh.
4. Visit eg `/posts` page
5. Login to the site and add the following to your `wp-config`
```
if ( ! empty( $_GET['rest_route'] ) ) {
	http_response_code( 404 );
	die('not found');
}
```
6. Ensure that you see the following error:
![rest_api_jetpack](https://github.com/Automattic/wp-calypso/assets/497103/ce4debbd-b251-4497-a65e-c39fc29ef14b)

### Test 2: XML-RPC errors
1. Apply D119857-code
2. Create a new WoA site.
3. Enable ssh.
4. Visit eg `/posts` page
5. Login to the site and add the following to your `wp-config`
```
if ( ! empty( $_GET['for'] ) && $_GET['for'] === 'jetpack' ) {
	http_response_code( 404 );
	die('not found');
}
```
6. Ensure that you see the following error:
![xml-rpc_jetpack](https://github.com/Automattic/wp-calypso/assets/497103/957f6cb4-f5a1-4517-9e61-bce68ff1fa89)



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

kozer committed:
Jetpack UX: Add xml and rest api error handling

danielbachhuber reviewed (commented):
Can you include test instructions? Also, should we link to the new support doc in these messages?

danielbachhuber commented on client/components/jetpack/connection-health/index.tsx:
```diff
@@ -79,6 +81,34 @@ export const JetpackConnectionHealthBanner = ( { siteId }: Props ) => {
+				errorType={ errorType }
+				errorText={ translate( 'Jetpack encounters REST API connection issues.' ) }
+				noticeActionHref={ localizeUrl(
+					'https://wordpress.com/support/why-is-my-site-down/#theres-an-issue-with-your-sites-jetpack-connection'
```
I created an issue to add an explanation to Resolve Jetpack Errors: https://github.com/Automattic/en.support-docs-content/issues/2080

kozer replied:
Should I create a ticket to revisit this and update the URL after the new docs are in place?

kozer committed:
Update: Add handling for XML-RPC error.

Our current implementation was lacking support when the site had only XML-RPC errors.
Changed the current logic everywhere it was used to use the new one.

sejas commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js:
```diff
@@ -19,3 +23,53 @@ export default function isJetpackConnectionProblem( state, siteId ) {
+/**
+ * React HOC to check if the current site has possible Jetpack connection problem.
+ */
+export const withJetpackConnectionProblem = ( Wrapped ) => ( props ) => {
+	const siteId = useSelector( ( state ) => getSelectedSiteId( state ) );
+	const isPossibleConnectionProblem = useIsJetpackConnectionProblem( siteId );
+	return (
+		<Wrapped { ...props } isPossibleJetpackConnectionProblem={ isPossibleConnectionProblem } />
+	);
+};
```
Nice HOC.

sejas commented on client/state/jetpack-connection-health/selectors/is-jetpack-connection-problem.js:
```diff
@@ -19,3 +23,53 @@ export default function isJetpackConnectionProblem( state, siteId ) {
+			}
+		};
+
+		/**
+		 * This event listener is used to detect if the site is inaccessible.
+		 * Among others it's used as a login in `wpcom-proxy-request` package,
+		 * and the idea is borrowed from there to detect the site inaccessibility.
+		 */
+		window.addEventListener( 'message', onMessage );
+
+		return () => {
+			window.removeEventListener( 'message', onMessage );
+		};
+	}, [ isPossibleConnectionProblem, siteId, dispatch ] );
```
I was trying to find other alternatives to this global listener, but I cannot imagine another way to do it.

kozer replied:
Yes, I searched all Jetpack reducer stuff before I ended up with that solution. I ended up enabling debug and noticed that the only consistent thing was the `wpcom-proxy-request` package. 
Searching a bit realized that the messages I saw were introduced using the same way I followed here.

sejas reviewed (approved):
I tested this PR along with D119857-code and it works as described.

The only two things I noticed that are not related to this PR are:
1. The jetpack error is cached and keeps showing me the error even after I've fixed it.
  - Between tests I had to change the health Jeptack cahce key on WPcom.
2. Sometimes when accessing to the site url I'm redirected to the global version
   - Example: `http://calypso.localhost:3000/posts/20230830testingjetpackresterror.wordpress.com` -> `http://calypso.localhost:3000/posts`

## Screenshots:

Before this PR:
<img width="1270" alt="Screenshot 2023-08-30 at 14 07 11" src="https://github.com/Automattic/wp-calypso/assets/779993/03897eaa-a290-486d-a51b-761c5519cfe6">



After this PR REST:

<img width="1276" alt="Screenshot 2023-08-30 at 14 13 27" src="https://github.com/Automattic/wp-calypso/assets/779993/dbb17796-f18f-4334-b81a-15907f5afc78">

After this PR XML-RPC:

<img width="1278" alt="Screenshot 2023-08-30 at 14 08 59" src="https://github.com/Automattic/wp-calypso/assets/779993/8a7c54d2-fa6a-4fc2-9675-a1ddf9b3c869">

danielbachhuber commented:
> > Sometimes when accessing to the site URL I'm redirected to the global version
> 
> This is something that is happening to me a lot lately, and I thought it was something wrong with the local environment or something.

The Redux cache is purged randomly when developing locally.

kozer committed:
Merge branch 'trunk' into add/jetpack_xml_rest_connection_errors

kozer committed:
Update: Add resolve jetpack errors page for now until new docs are ready

kozer committed:
Merge branch 'trunk' into add/jetpack_xml_rest_connection_errors

kozer committed:
Update: Check if window is undefined in an effort to make tests pass

kozer committed:
Update: Try to fix unit test errors, update types

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/8829018

Thank you @kozer for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/8829018).
